spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```diff
@@ -0,0 +1,319 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.functions;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.StandardCharsets;
+import org.apache.iceberg.spark.SparkSchemaUtil;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.util.ByteBuffers;
+import org.apache.iceberg.util.UnicodeUtil;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.catalog.functions.BoundFunction;
+import org.apache.spark.sql.connector.catalog.functions.ScalarFunction;
+import org.apache.spark.sql.connector.catalog.functions.UnboundFunction;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Decimal;
+import org.apache.spark.sql.types.DecimalType;
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * Implementation of {@link UnboundFunction} that matches the {@code truncate} transformation.
+ * The various implementations are registered with the {@link org.apache.iceberg.spark.SparkCatalog}
+ * such that the function can be used as {@code truncate(col, width)} or {@code truncate(col, 2)}.
+ */
+public class TruncateFunction implements UnboundFunction {
+
+  public TruncateFunction() {
+
+  }
+
+  private static void validateSourceTypeUsingSparkTypes(DataType dt) {
+    if (!DataTypes.IntegerType.acceptsType(dt) && !DataTypes.LongType.acceptsType(dt) &&
+        !DataTypes.StringType.acceptsType(dt) && !DataTypes.BinaryType.acceptsType(dt) &&
+        !(dt instanceof DecimalType)) {
+      throw new UnsupportedOperationException(
+          "Cannot truncate input type. Expected one of [IntegerType, LongType, DecimalType, StringType, BinaryType], " +
+              "but found " + dt.catalogString());
+    }
+  }
+
+  private static void validateTruncationWidthType(DataType widthType) {
+    if (!(widthType instanceof IntegerType ||
+        !(widthType instanceof LongType && DataTypes.IntegerType.acceptsType(widthType)))) {
+      throw new UnsupportedOperationException("Expected truncation width to be an integer type");
+    }
+  }
+
+  @Override
+  public BoundFunction bind(StructType inputType) {
+    if (inputType.fields().length != 2) {
+      throw new UnsupportedOperationException(
+          String.format("Invalid input type. Expected 2 fields but found %s", inputType.fields().length));
+    }
+
+    StructField fieldToTruncate = inputType.apply(0);
+    StructField truncationWidth = inputType.apply(1);
```

Comment:
The Spark [built-in `bucket` function](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L4976-L4978) makes the width the first argument. I think we should probably stick with that convention for `bucket` and `truncate`.

Reply:
Changed width to be the first argument.